wrc/codegen/merger.py

```diff
@@ -0,0 +1,179 @@
+from wrc.sema.ast import Article, Section, TableOfContent, Guideline, split_rule_number
+from copy import deepcopy
+
+NOTES_INDEX = 0
+TOC_INDEX = 1
+FIRST_ARTICLE_INDEX = 2
+
+
+class BadFormatError(Exception):
+    def __init__(self, message: str):
+        self.message = message + '. Please report this issue to the developer.'
+        super().__init__(message)
+
+
+def article_number_is_lower(num1: str, num2: str) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    This is necessary because:
+    * We need to compare article numbers like '1' and 'A'.
+    * '2' > '10' == True, which is a problem.
+    """
+    is_digit_1 = num1.isdigit()
+    is_digit_2 = num2.isdigit()
+    if is_digit_1 and is_digit_2:
+        return int(num1) < int(num2)
+    elif is_digit_1 and not is_digit_2:
+        # Only num2 is a letter.
+        return True
+    elif not is_digit_1 and not is_digit_2:
+        return num1 < num2
+
+    return False
+
+
+def reg_number_is_lower(num1: list, num2: list) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    We need this for the same reasons we need compare_article_numbers().
+    """
```

Comment:
```python3
def reg_number_is_lower(num1: list, num2: list) -> bool:
    """
    Returns True if num1 < num2, otherwise returns False.
    Note: this function is always used to compare regs that belong to the same article, so the order of numbers
    and letters is always the same for both num1 and num2.

    We want the following examples to evaluate to True:
    * [1, 'a'] < [1, 'a', 1] < [1, 'a', 2, 'b'] < [1, 'c']
    * ['B', 2] < ['B', 2, 'd'] < ['B', 2, 'd', 1] < ['B', 3]
    """
    for n1, n2 in zip(num1, num2):
        if n1 < n2:
            return True
        elif n1 > n2:
            return False

    # zip() includes the items up to the length of the shorter list.
    # We get here if the lengths are different but all the items included by zip are equal.
    # For example, if we are comparing: [2, 'a'] < [2, 'a', 1]. The for loop compares up to [2, 'a'], and as it does not
    # return from there, we just check if num1 had a shorter list than num2.
    return len(num1) < len(num2)
```